javaslang/generator/Generator.scala

```diff
@@ -1639,36 +1639,16 @@ def generateMainClasses(): Unit = {
                   } else {
                       ${val mappingFunction = (checked, i) match {
                           case (true, 0) => s"() -> $Try.of(this::apply).get()"
-                          case (true, 1) => s"t -> $Try.of(() -> this.apply(t)).get()"
-                          case (true, _) => s"t -> $Try.of(() -> tupled.apply(t)).get()"
+                          case (true, _) => s"t -> $Try.of(() -> apply($params)).get()"
                           case (false, 0) => s"this::apply"
-                          case (false, 1) => s"this"
-                          case (false, _) => s"tupled"
-                        }
-                        val forNull = (checked, i) match {
-                          case (true, 1) => s"$Try.of(() -> apply(null))::get"
-                          case (false, 1) => s"() -> apply(null)"
-                          case _ => null
+                          case (false, _) => s"tupled()"
                         }
                         if (i == 0) xs"""
                           return ($className$fullGenerics & Memoized) Lazy.of($mappingFunction)::get;
-                        """ else if (i == 1) xs"""
-                          final Object lock = new Object();
-                          final ${im.getType("java.util.Map")}<$generics, R> cache = new ${im.getType("java.util.HashMap")}<>();
-                          return ($className$fullGenerics & Memoized) t1 -> {
-                              synchronized (lock) {
-                                  return cache.computeIfAbsent(t1, $mappingFunction);
-                              }
-                          };
                         """ else xs"""
-                          final Object lock = new Object();
                           final ${im.getType("java.util.Map")}<Tuple$i<$generics>, R> cache = new ${im.getType("java.util.HashMap")}<>();
```

Comment:
Yes, that's a concern, as weak references can be eliminated at any time. The problem is even more problematic for `FunctionN` where `N ≥ 2`, as the key (tuple) is built internally, immediately discarding any other reference to it.

The above problem with the GC prematurely dropping the data could be eliminated by using _soft references_; for details, [see this](http://stackoverflow.com/questions/299659/what-is-the-difference-between-a-soft-reference-and-a-weak-reference-in-java):

> SoftReferences aren't required to behave any differently than  WeakReferences, but in practice softly reachable objects are generally retained as long as memory is in plentiful supply. This makes them an excellent foundation for a cache, such as the image cache described above, since you can let the garbage collector worry about both how reachable the objects are (a strongly reachable object will never be removed from the cache) and how badly it needs the memory they are consuming.